### ISSUE TITLE
Enable rust-analyzer optional cargo features

### ISSUE THREAD

**Issue by EnigmaCurry** [brotzeit/rustic#529]
I am playing around with [Leptos](https://github.com/leptos-rs/) which lets you write isomorphic Rust code that runs both on the server and in the web-browser. It uses code blocks like this to run code that needs to execute only in one environment or the other:

```
use cfg_if::cfg_if;

cfg_if! {
if #[cfg(feature = "ssr")] {
//   nb:  "ssr" means exclusive server side code!
     ... 
} else {
// run this on the client only 
}

// run anything else in the file in both environments
...
```

The problem is that rust-analyzer does not consider any of the code nested inside the "ssr" feature check, and it does not report on any coding mistakes or warnings inside the block.

According to https://github.com/rust-lang/rust-analyzer/issues/14045 rust-analyzer can only support ONE path through this block, but thats better than nothing. So the ssr feature can be enabled by setting the `rust-analyzer.cargo.features` setting by the lsp *client*, and this should enable rust-anaylzer to properly check the code inside the ssr block (at least).

So how do I set the `rust-analyzer.cargo.features` setting via rustic in Emacs? I found these likely variables and set them accordingly, but to no avail:

```
(custom-set-variables
 ;; custom-set-variables was added by Custom.
 '(flycheck-rust-features '("ssr"))
 '(lsp-rust-all-features t)
 '(lsp-rust-features ["ssr"]))
```

## Environment

```
$ emacs --version
GNU Emacs 29.1
$ rustc --version
`rustc` 1.74.0-nightly (ef85656a1 2023-08-21)
$ rust-analyzer --version
rust-analyzer 1.74.0-nightly (ef85656 2023-08-21)
```

rustic version: 3.4

**Comment by upbqdn** [comment]
> I found these likely variables and set them accordingly, but to no avail: [...]

I tried setting the same variables, but I still can't get error lints for code behind features.

**Comment by psibi** [comment]
If you would have to configure rust-analyzer, that should be done via lsp-mode. Eg: https://github.com/emacs-lsp/lsp-mode/pull/4257

**Comment by psibi** [comment]
When something like this happens, I try to check it's behavior with VSCode and see the language server logs and try to compare it with the lsp-mode interaction to see what is missing in lsp-mode.

**Comment by upbqdn** [comment]
> compare it with the lsp-mode interaction 

That would help. Is there a way I can actually see the communication between lsp-mode and rust-analyzer?

**Comment by shaneikennedy** [comment]
<img width="588" height="193" alt="Image" src="https://github.com/user-attachments/assets/76559aad-636e-4568-ba74-60d1206535be" />

I'm seeing this as well with eglot, has anyone had any luck? this makes writting a library with features really difficult in emacs

**Comment by shaneikennedy** [comment]
Found a solution, a rust-analyzer.toml file in the root of the project that enables all features
```toml
[cargo]
features = "all"

[check]
features = "all"
command = "check"
```

**Comment by Davoodeh** [comment]
TL;DR: Seems only `rust-analyzer.toml` in user configurations (`~/.config/rust-analyzer/rust-analyzer.toml`) method works as of 2026.

I know this is old but since the issue is still open and still comes up on search engines I thought I might add a little note here. Since `shaneikennedy` asked if anybody tried their luck with `eglot`.

I'm using
- doomemacs/doomemacs@57818a6
- Emacs 30.2-2 from Archlinux repositories
- `rust-analyzer` 20260420-1 from Archlinux repositories and also from `1.93.1` toolchains using `rustup` (I'm not entirely sure which one is preferred with my configuration).

After seeing different configurations and especially [this one by `ia0` on joaotavora/eglot (1214)](https://github.com/joaotavora/eglot/discussions/1214) I settled for the minimal configuration below to test the features:

```elisp
(setq rustic-analyzer-command
      '("rust-analyzer" :initializationOptions
        (:overrideCommand ("leptosfmt" "--stdin" "--rustfmt")
         :cargo (:features "all")
         :check (:features "all"
                 :command "clippy"))))
```

Equal to the following `rust-analyzer.toml`:
```toml
[rustfmt]
overrideCommand = ["leptosfmt", "--stdin", "--rustfmt"]

[cargo]
features = "all"

[check]
features = "all"
command = "clippy"
```

The confusing part for me is that some features work and some don't, under seemingly random conditions that I try to describe below.

The goal is for the following code to throw two compile errors:

```rust
#[cfg(feature = "ssr")]
compile_error!("ssr worked!");

#[cfg(feature = "hydrate")]
compile_error!("hydrate worked!");
```

So here is the different scenarios I tried and the results I got:

1. Unlike what he kindly suggested, having `rust-analyzer.toml` in the root of the project does not work (fully).
The `rustfmt.overrideCommand` will use `leptosfmt` judging by the fact that `view!` macros are formatted. But `features` keys do not work.
2. Having `rust-analyzer.toml` in the root of the project AND the Elisp configuration behaves like the first scenario (which had no Elisp modifications). Although it is worth mentioning that I am sure that the Elisp code is applied since providing invalid keys `:unkownKey` in the `plist` will throw `[eglot] ... unknown slot ...` errors and `eglot-server-programs` reflects this.
3. Having only the Elisp configuration seems to do absolutely nothing even though it is established that it gets applied correctly.
4. Having the `rust-analyzer.toml` put in `~/.config/rust-analyzer/` seems to work. However `eglot` throws the following errors in the messages (even though it works!)

```
[eglot] Server reports (type=2): invalid config values:
cargo/features: unexpected field
;
check/command: unexpected field
;
check/features: unexpected field
;
rustfmt/overrideCommand: unexpected field
;
```

See also [rust-analyzer/docs/book/src/configuration.md](https://github.com/rust-lang/rust-analyzer/blob/master/docs/book/src/configuration.md) and [rust-analyzer/docs/book/src/configuration_generated.md](https://github.com/rust-lang/rust-analyzer/blob/master/docs/book/src/configuration_generated.md) (last checked on rust-lang/rust-analyzer@f8018fc).

**Comment by upbqdn** [comment]
FWIW, this is my eglot config for rust-analyzer in Doom, which seems to work:

```elisp
(setq eglot-workspace-configuration
        '((:rust-analyzer .
           (:checkOnSave t
            :diagnostics (:disabled ["inactive-code"])
            :cfg         (:extra ["zcash_unstable"
                                  "zcash_unstable=\"zfuture\""
                                  "feature=\"nu7\""])
            :inlayHints  (:bindingModeHints (:enable t)
                          :lifetimeElisionHints (:enable t
                                                 :useParameterNames t))
            :check       (:command "clippy")
            :cargo       (:allFeatures t
                          :buildScripts (:enable t))
            :procMacro   (:enable t)))))
```
